base/util/PKICertImport.bash

```diff
@@ -65,6 +102,105 @@ function PKICertImport() {
+    function __secure_mktmp() {
+        local tmpdir="$TMPDIR"
+        local ret=0
+
+        # Prefer /dev/shm over /tmp: /dev/shm is less frequently backed by
+        # a physical disk than /tmp. However, if TEMPDIR is explicitly set,
+        # respect it.
+        if [ "x$tmpdir" == "x" ] && [ -d "/dev/shm" ]; then
+            tmpdir="/dev/shm"
+        elif [ "x$tmpdir" == "x" ] && [ -d "/tmp" ]; then
+            tmpdir="/tmp"
+        elif [ "x$tmpdir" == "x" ]; then
+            tmpdir="$HOME"
+        fi
+
+        TMPBASE="$(mktemp --directory --tmpdir="$tmpdir" 2>&1)"
+        ret="$?"
+
+        if (( ret != 0 )); then
+            __e "Return from mktemp was non-zero: $ret"
+            __e "$TMPBASE"
+            __e "Perhaps specify TMPDIR in the environment?"
+            exit 1
+        elif [ ! -d "$TMPBASE" ]; then
+            __e "Return from mktemp was zero but invalid directory:"
+            __e "$TMPBASE"
+            __e "Perhaps specify TMPDIR in the environment?"
+            exit 1
+        fi
+
+        # We've validated that TMPBASE is now a valid directory. Since
+        # we created it, we have ownership. Restrict access to only this
+        # user as the original NSS DB might have private keys which we want
+        # to keep secure when copying.
+
+        local user=""
+        local group=""
+
+        # Acquire current username.
+        user="$(id --user --name 2>&1)"
+        ret=$?
+        if (( ret != 0 )); then
+            __e "id exited with non-zero result: $ret"
+            __e "Unable to get current user's name."
+            __secure_rmtmp
+            exit 1
+        fi
+
+        # Acquire current primary group.
+        group="$(id --group --name 2>&1)"
+        ret=$?
+        if (( ret != 0 )); then
+            __e "id exited with non-zero result: $ret"
+            __e "Unable to get current user's name."
+            __secure_rmtmp
+            exit 1
+        fi
+
+        # Restrict permissions from
+        chown "$user:$group" -R "$TMPBASE"
+        ret=$?
+        if (( ret != 0 )); then
+            __e "Return from chown on $TMPBASE was non-zero: $ret"
+            __secure_rmtmp
+            exit 1
+        fi
+
+        chmod 700 -R "$TMPBASE"
+        ret=$?
+        if (( ret != 0 )); then
+            __e "Return from chmod on $TMPBASE was non-zero: $ret"
+            __secure_rmtmp
+            exit 1
+        fi
+
+        # Before continuing, save certificate to temporary location. This
+        # ensures we don't use accidentally use it.
+        cp "$CERT_PATH" "$TMPBASE/cert_path.orig"
+        CERT_PATH="$TMPBASE/cert_path.orig"
+
+        return 0
+    }
+
+    ## __secure_rmtmp removes the temporary directory if present. If the shred
+    ## command is present, removes files with shred.
+    function __secure_rmtmp() {
+        if [ -d "$TMPBASE" ]; then
+            if command -v shred >/dev/null 2>&1; then
```

Comment:
I'm not sure if this is strictly necessary. It probably only really matters when we're backed by a physical disk... Thoughts?